searx/shared/__init__.py

```diff
@@ -4,33 +4,47 @@
 
 logger = logging.getLogger('searx.shared')
 
+__all__ = ["storage", "schedule"]
+
+
 try:
-    import uwsgi
-except:
-    # no uwsgi
-    from .shared_simple import SimpleSharedDict as SharedDict, schedule
+    # First: try to use Redis
+    from .redisdb import client
+
+    client().ping()
+    from .shared_redis import RedisCacheSharedDict as SharedDict, schedule
 
-    logger.info('Use shared_simple implementation')
-else:
+    logger.info('Use shared_redis implementation')
+except Exception as e:
+    # Second: try to use uwsgi
     try:
-        uwsgi.cache_update('dummy', b'dummy')
-        if uwsgi.cache_get('dummy') != b'dummy':
-            raise Exception()
+        import uwsgi
     except:
-        # uwsgi.ini configuration problem: disable all scheduling
-        logger.error(
-            'uwsgi.ini configuration error, add this line to your uwsgi.ini\n'
-            'cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1'
-        )
-        from .shared_simple import SimpleSharedDict as SharedDict
-
-        def schedule(delay, func, *args):
-            return False
+        # Third : fall back to shared_simple
+        from .shared_simple import SimpleSharedDict as SharedDict, schedule
 
+        logger.info('Use shared_simple implementation')
     else:
-        # uwsgi
-        from .shared_uwsgi import UwsgiCacheSharedDict as SharedDict, schedule
-
-        logger.info('Use shared_uwsgi implementation')
+        # Make sure uwsgi is okay
+        try:
+            uwsgi.cache_update('dummy', b'dummy')
+            if uwsgi.cache_get('dummy') != b'dummy':
+                raise Exception()  # pylint: disable=raise-missing-from
+        except:
+            # there is exception on a get/set test: disable all scheduling
+            logger.exception(
+                'uwsgi.ini configuration error, add this line to your uwsgi.ini\n'
+                'cache2 = name=searxngcache,items=2000,blocks=2000,blocksize=4096,bitmap=1\n'
+            )
+            from .shared_simple import SimpleSharedDict as SharedDict
+
+            def schedule(delay, func, *args):
+                return False
```

Comment:
Why not using (import) `.shared_simple.schedule` here? In line 24 `schedule` is imported but not here .. what is the difference?

Reply:
the scheduler is disabled because:
* the app is managed by uwsgi (we can import `uwsgi`), so there are multiple processes.
* with the simple implementation, the scheduled task is going to run on each process.

So the best thing to do is:
* either stop the app and ask the admin to fix the issue.
* or just disable the scheduler because it is only required by the checker.

Reply:
> So the best thing to do is:

What do think about my comment above [.. to me its very complicated .. isn't it better we drop the uWSGI support and install a new process for the scheduler. This will simplify the code a lot.](https://github.com/searxng/searxng/pull/1513#pullrequestreview-1064421701)

Reply:
> the Python packages python-crontab and scheduler don't solve the requirement to run the task once per instance (not per worker). It is implemented in this PR in 110 lines of code.

May be it is not clear what I have in mind about .. let me explain:

From the architecture point of view, I think it is wrong to have an *in-process* in the WEB-App where you have to be careful not to mix it with the workers (threads or whatever) of the WEB-App:

We have a WEB-App and we have services. For the (temporary) persistence we use redis and for the scheduler we also need a service.

> in the end I'm afraid it will move the complexity from us to the admins.

It's just a question of how we orchestrate these services;

- redis is e.g. run by the operating system (systemd),
- our WEB app is also managed by systemd and
- the scheduler should also be managed  by the operating system.

The orchestration using the crontab is really a bit complicated, but it is also quite simple using systemd and define `WantedBy=` units.

By example: we have one central systemd unit and two `WantedBy=` units, one is the WEB-App (wsgi) and the other is a scheduler process.

![image](https://user-images.githubusercontent.com/554536/183602240-608f51d4-3cac-4e2a-8c21-2680e38d5ddc.png)

The `WantedBy=` units can be controlled via the central service (start/stop), additionally the single services can be switched off or disabled via systemd.

I think this is an *extension* of the possibilities without the admin having to pay more attention than before.

Here is a detail example: https://alesnosek.com/blog/2016/12/04/controlling-a-multi-service-application-with-systemd/

Reply:
> The orchestration using the crontab is really a bit complicated, but it is also quite simple using systemd and define WantedBy= units.

Some OS do not have systemd like [devuan](https://www.devuan.org/) or MacOS.

> I think it is wrong to have an in-process in the WEB-App 

Yes, I agree, that's why there is [Celery](https://docs.celeryq.dev/en/stable/) for example in the Python ecosystem.

The purpose is to remove the uwsgi requirement since multiple users reports issues with it.

This PR prepare a replacement by a Redis implementation of 110 lines that are not by the book for sure, but do the job.

Reply:
> Some OS do not have systemd like [devuan](https://www.devuan.org/) or MacOS.

.. and MS-Win.  IMO this is not an argument since we do not give any support on deployment on this OS .. uWSGI support by example is also unknown to us.

> Yes, I agree, that's why there is [Celery](https://docs.celeryq.dev/en/stable/) for example in the Python ecosystem.

Do we really need to implement a task queues?  The only argument - I see - could be to be independent from OS .. but is this really a job of the SearXNG app?

I would like to avoid that we always reach for new tools only because we are convinced that SearXNG has to take over tasks that are actually done by an OS.

> This PR prepare a replacement by a Redis implementation of 110 lines

If you think this PR is a good first solution that fixes an existing issue, I don't want to block the PR from a merge and we can implement a different scheduler in another PR.

But to me it becomes harder and harder to understand .. besides the complexity of being in a uWSGI environment or not, this PR adds the complexity of additional redis scripts. 

TBH I do understand only the half of all / the ` RedisCacheSharedDict` is simple, but the scheduler part is hard for me.

Anyway this might only be my problem .. if you think this PR is a valid solution, lets merge.